src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -386,404 +210,472 @@ private StateResult Data()
                     new SourceSpan(CurrentStart, contentLength: 1 /* end of file */)));
-
-            return Transition(CSharpTokenizerState.Data, EndToken(SyntaxKind.CSharpComment));
         }
-        if (CurrentCharacter == '*')
+
+        // TODO: Handle preprocessor directives
+        var tokenType = trivia.Kind() switch
         {
-            TakeCurrent();
-            if (CurrentCharacter == '/')
-            {
-                TakeCurrent();
-                return Transition(CSharpTokenizerState.Data, EndToken(SyntaxKind.CSharpComment));
-            }
-        }
-        return Stay();
+            CSharpSyntaxKind.WhitespaceTrivia => SyntaxKind.Whitespace,
+            CSharpSyntaxKind.EndOfLineTrivia => SyntaxKind.NewLine,
+            CSharpSyntaxKind.SingleLineCommentTrivia or CSharpSyntaxKind.MultiLineCommentTrivia or CSharpSyntaxKind.MultiLineDocumentationCommentTrivia or CSharpSyntaxKind.SingleLineDocumentationCommentTrivia => SyntaxKind.CSharpComment,
+            _ => throw new InvalidOperationException("Unexpected trivia kind."),
+        };
+
+        return Stay(EndToken(tokenType));
     }
 
-    // CSharp Spec §2.3.2
-    private StateResult SingleLineComment()
+    private StateResult OnRazorCommentStar()
     {
-        TakeUntil(c => SyntaxFacts.IsNewLine(c));
-        return Stay(EndToken(SyntaxKind.CSharpComment));
+        AssertCurrent('*');
+        TakeCurrent();
+
+        return Transition(
+            RoslynCSharpTokenizerState.RazorCommentBody,
+            EndToken(SyntaxKind.RazorCommentStar));
     }
 
     // CSharp Spec §2.4.4
     private StateResult NumericLiteral()
     {
-        if (TakeAll("0x", caseSensitive: true))
+        var curPosition = Source.Position;
+        var result = GetNextResult(NextResultType.Token);
+        var csharpToken = result.Token;
+        // Don't include trailing trivia; we handle that differently than Roslyn
+        var finalPosition = curPosition + csharpToken.Span.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
-            return HexLiteral();
+            TakeCurrent();
         }
-        else
+
+        Buffer.Clear();
+        _currentCSharpTokenTriviaEnumerator = (csharpToken.TrailingTrivia.GetEnumerator(), isLeading: false);
+        return Transition(RoslynCSharpTokenizerState.TriviaForCSharpToken, EndToken(csharpToken.Text, SyntaxKind.NumericLiteral));
+    }
+
+    private StateResult Identifier()
+    {
+        var curPosition = Source.Position;
+        var result = GetNextResult(NextResultType.Token);
+        var csharpToken = result.Token;
+        // Don't include trailing trivia; we handle that differently than Roslyn
+        var finalPosition = curPosition + csharpToken.Span.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
-            return DecimalLiteral();
+            TakeCurrent();
         }
```

Comment:
This code appears to be duplicated in multiple places in the class, could this be refactored into a private method that can be called from multiple call sites?